common/util_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"testing"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"github.com/uber-common/bark"
+)
+
+var baseLog = bark.NewLoggerFromLogrus(logrus.StandardLogger())
+
+type UtilSuite struct {
+	*require.Assertions // override suite.Suite.Assertions with require.Assertions; this means that s.NotNil(nil) will stop the test, not merely log an error
+	suite.Suite
+}
+
+func TestUtilSuite(t *testing.T) {
+	suite.Run(t, new(UtilSuite))
+}
+
+func (s *UtilSuite) SetupTest() {
+	s.Assertions = require.New(s.T()) // Have to define our overridden assertions in the test setup. If we did it earlier, s.T() will return nil
+}
+
+var logFnInvocations int
+
+func logFn() bark.Logger {
+	logFnInvocations++
+	return baseLog.WithField(TagModule, `util_test`)
+}
+
+func (s *UtilSuite) TestOverrideValueByPrefixNil() {
+	logFnInvocations = 0
+	x := OverrideValueByPrefix(logFn, `/foo/bar`, nil, 42, `TestOverrideValueByPrefixNil`)
+	s.EqualValues(x, 42)
+	s.Equal(logFnInvocations, 0)
+
+	// Verify that we log an error every time that we pass invalid rules
+	for i := 1; i < 4; i++ {
+		x = OverrideValueByPrefix(logFn, `/foo/bar`, make([]string, 5), 42, `2TestOverrideValueByPrefixNil`)
+		s.EqualValues(x, 42)
+		s.Equal(logFnInvocations, i*5)
+	}
+
+	badRules := []string{`foo`, `/=`, `/foo=`, `/fo=bar`, `=baz`}
+
+	logFnInvocations = 0
+	for i := 1; i < 4; i++ {
+		x = OverrideValueByPrefix(logFn, `/foo/bar`, badRules, 42, `3TestOverrideValueByPrefixNil`)
+		s.EqualValues(x, 42)
+		s.Equal(logFnInvocations, i*5)
+	}
+}
+
+func (s *UtilSuite) TestOverrideValueByPrefixNormal() {
```

Comment:
Would be cool to have one of these tests run the rule evaluations concurrently i.e. spin go routines inside your for loop. That would have caught the concurrent unsafe access to map.